src/graph/sampling/metapath_randomwalk.h

```diff
@@ -0,0 +1,147 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file graph/sampler/generic_randomwalk_cpu.h
+ * \brief DGL sampler - templated implementation definition of random walks on CPU
+ */
+
+#ifndef DGL_GRAPH_SAMPLING_METAPATH_RANDOMWALK_H_
+#define DGL_GRAPH_SAMPLING_METAPATH_RANDOMWALK_H_
+
+#include <dgl/array.h>
+#include <dgl/base_heterograph.h>
+#include <dgl/random.h>
+#include <utility>
+#include <vector>
+#include "randomwalks.h"
+#include "randomwalks_cpu.h"
+
+namespace dgl {
+
+using namespace dgl::runtime;
+using namespace dgl::aten;
+
+namespace sampling {
+
+namespace impl {
+
+namespace {
+
+// bool WhetherToTerminate(
+//     void *node_ids_generated_so_far,
+//     dgl_id_t last_node_id_generated,
+//     int64_t number_of_nodes_generated_so_far)
+using TerminatePredicate = std::function<bool(void *, dgl_id_t, int64_t)>;
+
+/*!
+ * \brief Select one successor of metapath-based random walk, given the path generated
+ * so far.
+ *
+ * \param data The path generated so far, of type \c IdxType.
+ * \param curr The last node ID generated.
+ * \param len The number of nodes generated so far.  Note that the seed node is always
+ *            included as \c data[0], and the successors start from \c data[1].
+ *
+ * \param edges_by_type Vector of results from \c GetAdj() by edge type.
+ * \param metapath_data Edge types of given metapath.
+ * \param prob Transition probability per edge type.
+ * \param terminate Predicate for terminating the current random walk path.
+ *
+ * \return A pair of ID of next successor (-1 if not exist), as well as whether to terminate.
+ */
+template<DLDeviceType XPU, typename IdxType>
+std::pair<dgl_id_t, bool> MetapathRandomWalkStep(
+    void *data,     // void * because std::function does not accept template typenames.
```

Comment:
Can't it be `IdxType*`?

Reply:
I'm afraid not, because the first argument of `TerminatePredicate` and `StepFunc` must be `void *`.

Reply:
Would [type alias](https://zh.cppreference.com/w/cpp/language/type_alias) help here?

Reply:
NVM.  The comment above was a mistake.